### PR TITLE
Pessimising moves and missing overrides

### PR DESCRIPTION
**Pull request by chfast**
Non functional change. Removes explicit and pessimising moves and adds missing override specifiers.

### CODE REVIEW

File: json_spirit/json_spirit_writer_template.h [hunk 1]
```diff
@@ -25,13 +25,9 @@ namespace json_spirit
         return 'A' - 10 + ch;
     }
 
-#pragma GCC diagnostic push
```
Comment (chriseth) [inline json_spirit/json_spirit_writer_template.h]: Was that change intended? This is an external library.

**Comment by chriseth** [comment]
Ah, nice change! :-)

File: json_spirit/json_spirit_writer_template.h (hunk 1 above)
Comment (chfast) [inline json_spirit/json_spirit_writer_template.h]: I know. But we added macros to suppress warning about unused typedef. The same must be done for clang. I think it is better to remove that unused typedef.
